plugin.py

```diff
@@ -49,26 +49,50 @@ def prepare(self):
         if os.path.isfile(self.module_output):
             os.remove(self.module_output)
 
-    # run inside gui-thread-safe context
-    def invoke(self, runner):
-        wx.CallAfter(runner)
-
+    # export board.dsn file from pcbnew
     def RunExport(self):
-        pcbnew.ExportSpecctraDSN(self.module_input)
+        ok = pcbnew.ExportSpecctraDSN(self.module_input)
+        if ok and os.path.isfile(self.module_input):
+            return True
+        else:
+            show_error("Failed to invoke pcbnew.ExportSpecctraDSN")
+            return False
         
+    # auto route by invoking FreeRouting.jar
     def RunRouter(self):
         command = [self.java_path, "-jar", self.module_path, "-de", self.module_input, "-s"]
-        process = subprocess.Popen(command)
-        process.wait(self.module_timeout)
+        try:
+            process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            output, error = process.communicate()
+            if process.returncode == 0:
+                return True
+            else:  
+                show_error("Java program failure: %d\n %s\n %s\n" % (process.returncode, output, error))
+                return False
+        except Exception as error:
+                show_error("Java process failure: %s" % str(error))
+                return False
 
+    # import generated board.ses file into pcbnew
     def RunImport(self):
-        pcbnew.ImportSpecctraSES(self.module_output)
+        ok = pcbnew.ImportSpecctraSES(self.module_output)
+        if ok and os.path.isfile(self.module_output):
+            return True
+        else:
+            show_error("Failed to invoke pcbnew.ImportSpecctraSES")
+            return False
+        
+    # invoke chain of dependent methods
+    def RunSteps(self):
+        return self.RunExport() and self.RunRouter() and self.RunImport()
 
+    # kicad plugin action entry
     def Run(self):
+        if not has_pcbnew_api():
+            show_error("Missing required python API:\n pcbnew.ExportSpecctraDSN\n pcbnew.ImportSpecctraSES\n")
```

Comment:
I would add a note that plugin requires development nightly build from 2019-05-05 or later.

Reply:
applied